combat-trainer.lic

```diff
@@ -1997,7 +2019,189 @@ class TrainerProcess
     end
     DRRoom.dead_npcs.size > dead_count
   end
+  
+  def visible_bodies
+    result = []
+    all_bodies = @constellations
+    case DRC.bput('observe sky', 'The following heavenly bodies are visible:', "That's a bit hard to do while inside")
+    when "That's a bit hard to do while inside"
+      echo('Must be outdoors to observe sky')
+      exit
+    end
+    until (line = get?) =~ /^Roundtime/i
+      result << all_bodies.find { |body| /\b#{body['name']}\b/i =~ line && line !~ /below the horizon/ }
+    end
+    result.compact.select { |data| data['circle'] <= DRStats.circle }
+  end
+  
+  def check_pools
+    pools = {
+      'lore' => 0,
+      'magic' => 0,
+      'survival' => 0,
+      'offensive combat' => 0,
+      'defensive combat' => 0,
+      'future events' => 0
+    }
+    pools_to_size = {
+      /You have no understanding of the celestial influences over/ => 0,
+      /You have a feeble understanding of the celestial influences over/ => 1,
+      /You have a weak understanding of the celestial influences over/ => 2,
+      /You have a fledgling understanding of the celestial influences over/ => 3,
+      /You have a modest understanding of the celestial influences over/ => 4,
+      /You have a decent understanding of the celestial influences over/ => 5,
+      /You have a significant understanding of the celestial influences over/ => 6,
+      /You have a potent understanding of the celestial influences over/ => 7,
+      /You have an insightful understanding of the celestial influences over/ => 8,
+      /You have a powerful understanding of the celestial influences over/ => 9,
+      /You have a complete understanding of the celestial influences over/ => 10
+    }
+    waitrt?
+    DRCT.retreat
+    bput('predict state all', 'roundtime')
+    lines = reget(50).reverse
+    lines.select! { |line| line =~ /celestial influences/ }
+    pools.select! { |name, _level| pools[name] = lines.find { |line| line =~ /#{name}/ } }
+    pools.each { |name, text| pools_to_size.each { |match_text, value| pools[name] = value if match_text =~ text } }
+    echo " pools: #{pools}" if UserVars.astrology_debug
+    waitrt?
+    pools
+  end 
+  
+  def predict_all(pools)
+    skillset_to_pool = {
+      'offensive combat' => @astrology_prediction_skills_offense,
+      'defensive combat' => @astrology_prediction_skills_defense,
+      'magic' => @astrology_prediction_skills_magic,
+      'survival' => @astrology_prediction_skills_survival,
+      'lore' => @astrology_prediction_skills_lore,
+      'future events' => 'future events'
+    }
+
+    pools.reject { |_skill, size| (size < @prediction_pool_target) }
+         .each_key do |skill|
+      break if DRSkill.getxp('Astrology') > 30
+      align(skillset_to_pool[skill])
+    end
+  end 
+
+  def check_heavens
+    if snap != [left_hand, right_hand]
```

Comment:
`snap` is not defined. I am not sure if this works. If it does great, but i think its a very confusing way to declare a variable